test/jit/test_complex.py

```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
         t2 = torch.tensor([0.4, 1.4j, 2.35])
 
         self.checkScript(fn, (t1, t2, 2))
+
+    def test_complex_math_ops(self):
+        inf = float("inf")
+        NaN = float("nan")
+        vals = ([inf, NaN, 0.0, 1.0, 2.2, -1.0, -0.0, -2.2, -inf, 1, 0, 2]
+                + [10.0 ** i for i in range(5)] + [-(10.0 ** i) for i in range(5)])
+        complex_vals = tuple(complex(x, y) for x, y in product(vals, vals))
+
+        def checkMath(func_name, num_args, ret_type="complex", debug=False, vals=None, args_type=None):
+            funcs_template = dedent('''
+            def func(a):
+                # type: {args_type} -> {ret_type}
+                return cmath.{func}({args})
+            ''')
+            if num_args == 1:
+                args = "a"
+            elif num_args == 2:
+                args = "a, b"
+            else:
+                raise RuntimeError("Test doesn't support more than 2 arguments")
+            if args_type is None:
+                args_type = "(complex)"
+            funcs_str = funcs_template.format(func=func_name, args=args, args_type=args_type, ret_type=ret_type)
+            scope = {}
+            execWrapper(funcs_str, globals(), scope)
+            cu = torch.jit.CompilationUnit(funcs_str)
+            f_script = cu.func
+            f = scope['func']
+
+            for a in complex_vals:
+                res_python = None
+                res_script = None
+                try:
+                    res_python = f(a)
+                except Exception as e:
+                    res_python = e
+                try:
+                    res_script = f_script(a)
+                except Exception as e:
+                    res_script = e
+                if debug:
+                    print("in: ", a)
+                    print("out: ", res_python, res_script)
+
+                if res_python != res_script:
+                    if isinstance(res_python, Exception):
+                        continue
+
+                    if type(res_python) == type(res_script):
+                        if isinstance(res_python, tuple) and (cmath.isnan(res_python[0]) == cmath.isnan(res_script[0])):
+                            continue
+                        if isinstance(res_python, complex) and cmath.isnan(res_python) and cmath.isnan(res_script):
+                            continue
+                    msg = ("Failed on {func_name} with input {a}. Python: {res_python}, Script: {res_script}"
+                           .format(func_name=func_name, a=a, res_python=res_python, res_script=res_script))
+                    self.assertEqual(res_python, res_script, msg=msg)
+
+        unary_ops = ['log', 'log10', 'sqrt', 'exp', 'sin', 'cos', 'asin', 'acos', 'atan', 'sinh', 'cosh',
+                     'tanh', 'asinh', 'acosh', 'atanh']
```

Comment:
Is there any way to generate this list based on the contents of `cmath` so that new ops are automatically tested?

Reply:
Mmm I am not sure, but I think it's ok to stay with this list since there aren't any unary ops not present in cmath and are supported for complex in pytorch.